### PR TITLE
skip tuples for `no_magic_numbers` rule

### PR DESCRIPTION
**Pull request by mildm8nnered**
Now ignores magic numbers in tuple assignments like

`let (httpStatusCodeErrorLowerBound, httpStatusCodeErrorUpperBound) = (400, 599)`

OSS checks look good to me.

Resolves https://github.com/realm/SwiftLint/issues/5305 by skipping magic numbers in tuple expressions altogether.

### CODE REVIEW

File: CHANGELOG.md
Comment (mildm8nnered) [inline CHANGELOG.md:53]: Fixed some trailing spaces while I was here

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
Comment (SimplyDanny) [inline Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift:78]: The variable names could be shorter.

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
Comment (SimplyDanny) [inline Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift:95]: That's unnecessary code actually.

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
Comment (SimplyDanny) [inline Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift:78]: What about the case `let a = (2, 3)`? Is it not supposed to be ignored by the rule?

Comment (mildm8nnered) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift]: So I wasn't intending to cover that case, but it would be easy to. Do you think we should?

Comment (SimplyDanny) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift]: I think so. `let (a, b) = (2, 3)` is only a special case of `let a = (2, 3)`. But with regards to magic numbers it should not make a difference.

File: CHANGELOG.md
Comment (SimplyDanny) [inline CHANGELOG.md:69]:
```suggestion
  are defined in a tuple like `let (a, b) = (5, 10)` or `let a = (2, 3).  
```

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
Comment (mildm8nnered) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift]: So much nicer

**Comment by SimplyDanny** [comment]
Looking at the OSS findings, there's another case I haven't thought of before. If there are more complicated expression in a tuple initialization like in `let a = (2 + 3, f(arg: 47))`, the rule won't trigger now. Is this expected? In `let a = f(4)` the rule would trigger on `4`, so not triggering in tuples seems like an inconsistency then.

**Comment by mildm8nnered** [comment]
> Looking at the OSS findings, there's another case I haven't thought of before. If there are more complicated expression in a tuple initialization like in `let a = (2 + 3, f(arg: 47))`, the rule won't trigger now. Is this expected? In `let a = f(4)` the rule would trigger on `4`, so not triggering in tuples seems like an inconsistency then.

So `let a = f(4)` does trigger, even though it's technically a tuple, because there were OSS samples like `let imageHeight = (width - ↓24)` that I wanted to still warn on.

I *thought* the OSS violations we ended up with looked ok, but `let a = (2 + 3, f(arg: 47))` does bother me. That would not violate anymore.

Maybe drop back to just allowing

`let (a, b) = (5, 10)`

That would still allow

`let (a, b) = (2 + 3, f(arg: 47))`

or we could go further and require that the tuple elements all be simple expressions.

**Comment by SimplyDanny** [comment]
> ... or we could go further and require that the tuple elements all be simple expressions.

I think that's the way to go. The rule currently triggers on `let a = 2 + 3` but not on `let a = 3`. It should be the same for tuples in my opinion. I'm open for better arguments though.

**Comment by mildm8nnered** [comment]
> > ... or we could go further and require that the tuple elements all be simple expressions.
> 
> I think that's the way to go. The rule currently triggers on `let a = 2 + 3` but not on `let a = 3`. It should be the same for tuples in my opinion. I'm open for better arguments though.

Sounds good to me -  `let (a, b) = (5, 10)` all good. And a nope for anything else, or complex expressions.

**Comment by SimplyDanny** [comment]
> > > ... or we could go further and require that the tuple elements all be simple expressions.
> >
> > I think that's the way to go. The rule currently triggers on `let a = 2 + 3` but not on `let a = 3`. It should be the same for tuples in my opinion. I'm open for better arguments though.
> 
> Sounds good to me -  `let (a, b) = (5, 10)` all good. And a nope for anything else, or complex expressions.

Just to be sure: Also `let a = (3, 2)` wouldn't trigger, right?

**Comment by mildm8nnered** [comment]
> > > > ... or we could go further and require that the tuple elements all be simple expressions.
> > > 
> > > 
> > > I think that's the way to go. The rule currently triggers on `let a = 2 + 3` but not on `let a = 3`. It should be the same for tuples in my opinion. I'm open for better arguments though.
> > 
> > 
> > Sounds good to me -  `let (a, b) = (5, 10)` all good. And a nope for anything else, or complex expressions.
> 
> Just to be sure: Also `let a = (3, 2)` wouldn't trigger, right?

ok - so I think I have this working. 

As long as every element of the RHS tuple is an Integer or Float, we will allow it through, so

`let a = (3, 2)` is fine, as is `let (a, b) = (3, 2)`

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift
Comment (SimplyDanny) [inline Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift:79]: Let's also have a test for labeled tuple elements and a tuple used somewhere else that triggers the rule.

After that, I promise to accept the change. 😇

Comment (mildm8nnered) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift]: So I added those examples, and also extended the allowed tuple elements to include string literals, as cases like `let notFound = (404, "Not Found")` seemed reasonable to skip as well.

Comment (mildm8nnered) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift]: But hold off on merging for right now @SimplyDanny  - I want to checkout the OSS results, which now seem to have suspiciously lower fixes compared to before

Comment (mildm8nnered) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift]: ok - that looks better. Should be good to go now once CI passes

Comment (SimplyDanny) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift]: The strings and booleans make me think again. We now have:

```swift
let a = (3, 4)      // ok
let a = (3, "a")    // ok 
let a = (3, b)      // triggers on 3
let a = (3, (4, 5)) // triggers on all numbers
let a = (true, 4)   // ok
let a = (3, 4 + 5)  // triggers on all numbers
```

Does this make sense? Is it consistent and logical? 🤔 We might need to come up with a better set of rules for tuples I'm afraid.

Comment (mildm8nnered) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift]: So I totally get your point, but actually, all of the above examples would have been violations before, and there were enough reasonably allowable fixed violations in the OSS tests, that I think the change will add some value for some users.

```
let a = (3, b)      // triggers on 3
let a = (3, (4, 5)) // triggers on all numbers
let a = (3, 4 + 5)  // triggers on all numbers
```

will still be violations, but in the OSS code, there were few or none of these.

Users of the above examples would not be put at any disadvantage by this change - they would just need to disable the rule, like they would today.

I could definitely see a case for allowing `let a = (3, (4, 5))`, but if someone wants that to pass, they can always request that we extend the coverage.

Comment (SimplyDanny) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoMagicNumbersRule.swift]: Fair enough. Let's have it merged then. Good job!

**Review by SimplyDanny — APPROVED** [review]

**Pull request merged**
